assessment/collectors/infoschema_collector.go

```diff
@@ -0,0 +1,195 @@
+	"github.com/GoogleCloudPlatform/spanner-migration-tool/assessment/sources/mysql"
+	"github.com/GoogleCloudPlatform/spanner-migration-tool/assessment/utils"
+	"github.com/GoogleCloudPlatform/spanner-migration-tool/common/constants"
+	"github.com/GoogleCloudPlatform/spanner-migration-tool/conversion"
+	"github.com/GoogleCloudPlatform/spanner-migration-tool/internal"
+	"github.com/GoogleCloudPlatform/spanner-migration-tool/logger"
+	"github.com/GoogleCloudPlatform/spanner-migration-tool/profiles"
+)
+
+type InfoSchemaCollector struct {
+	tables           []utils.TableAssessment
+	indexes          []utils.IndexAssessment
+	triggers         []utils.TriggerAssessment
+	storedProcedures []utils.StoredProcedureAssessment
+	conv             *internal.Conv
+}
+
+func (c InfoSchemaCollector) IsEmpty() bool {
+	if c.conv == nil && c.indexes == nil && c.tables == nil && c.storedProcedures == nil && c.triggers == nil {
+		return true
+	}
+	return false
+}
+
+func CreateInfoSchemaCollector(conv *internal.Conv, sourceProfile profiles.SourceProfile) (InfoSchemaCollector, error) {
+	logger.Log.Info("initializing infoschema collector")
+	dbIdentifier := utils.DbIdentifier{}
+	if sourceProfile.Conn.Ty == profiles.SourceProfileConnectionTypeMySQL {
+		dbIdentifier.DatabaseName = sourceProfile.Conn.Mysql.Db
+	}
+	tb := []utils.TableAssessment{}
+	var errString string
+	for _, table := range conv.SrcSchema {
+		columnAssessments := []utils.ColumnAssessment[any]{}
+		for _, column := range table.ColDefs {
+			columnAssessments = append(columnAssessments, utils.ColumnAssessment[any]{
+				Db:        dbIdentifier,
+				Name:      column.Name,
+				TableName: table.Name,
+				ColumnDef: column,
+			})
+		}
+		tb = append(tb, utils.TableAssessment{Name: table.Name, TableDef: table, ColumnAssessments: columnAssessments, Db: dbIdentifier})
+	}
+	infoSchema, err := getInfoSchema(sourceProfile)
+	if err != nil {
+		return InfoSchemaCollector{}, err
+	}
+	indCollector := []utils.IndexAssessment{}
+	for _, table := range conv.SrcSchema {
+		index, err := infoSchema.GetIndexInfo(dbIdentifier.DatabaseName, table.Name)
+		if err != nil {
+			errString = errString + fmt.Sprintf("\nError while scanning indexes: %v", err)
+		}
+		indCollector = append(indCollector, index...)
+	}
+	triggers, err := infoSchema.GetTriggerInfo(dbIdentifier.DatabaseName)
+	if err != nil {
+		errString = errString + fmt.Sprintf("\nError while scanning triggers: %v", err)
+	}
+	sps, err := infoSchema.GetStoredProcedureInfo(dbIdentifier.DatabaseName)
+	if err != nil {
+		errString = errString + fmt.Sprintf("\nError while scanning stored procedures: %v", err)
+	}
+	err = nil
+	if errString != "" {
+		err = fmt.Errorf(errString, "")
+	}
+	return InfoSchemaCollector{
+		tables:           tb,
+		indexes:          indCollector,
+		triggers:         triggers,
+		conv:             conv,
+		storedProcedures: sps,
+	}, err
+}
+
+func getInfoSchema(sourceProfile profiles.SourceProfile) (common.InfoSchema, error) {
```

Comment:
suggest this interface be made a little more generic - so that it can be used with other databases and maybe even dump files.